content/blog/2021/10/05/OptimizeYourPython.adoc

```diff
@@ -0,0 +1,438 @@
+
+Hard constraints are weighted against other hard constraints. Soft constraints are weighted too, against other soft constraints. Hard constraints always outweigh soft constraints, regardless of their respective weights.
+
+To calculate the score, create a constraint provider function:
+
+[source,python]
+----
+from optapy import constraint_provider, get_class
+from optapy.types import Joiners, HardSoftScore
+
+# Constraint Factory takes Java Classes, not Python Classes
+LessonClass = get_class(Lesson)
+RoomClass = get_class(Room)
+
+@constraint_provider
+def defineConstraints(constraintFactory):
+    return [
+        # Hard constraints
+        roomConflict(constraintFactory),
+        teacherConflict(constraintFactory),
+        studentGroupConflict(constraintFactory),
+        # Soft constraints are only implemented in the optapy-quickstarts code
+    ]
+
+def roomConflict(constraintFactory):
+    # A room can accommodate at most one lesson at the same time.
+    return constraintFactory \
+            .fromUniquePair(LessonClass,
+                [
+                    # ... in the same timeslot ..
+                    Joiners.equal(lambda lesson: lesson.timeslot),
+                    # ... in the same room ...
+                    Joiners.equal(lambda lesson: lesson.room)
+                ]) \
+            .penalize("Room conflict", HardSoftScore.ONE_HARD)
+
+
+def teacherConflict(constraintFactory):
+    # A teacher can teach at most one lesson at the same time.
+    return constraintFactory \
+                .fromUniquePair(LessonClass,
+                        [
+                            Joiners.equal(lambda lesson: lesson.timeslot),
+                            Joiners.equal(lambda lesson: lesson.teacher)
+                        ]) \
+                .penalize("Teacher conflict", HardSoftScore.ONE_HARD)
+
+def studentGroupConflict(constraintFactory):
+    # A student can attend at most one lesson at the same time.
+    return constraintFactory \
+            .fromUniquePair(LessonClass,
+                [
+                    Joiners.equal(lambda lesson: lesson.timeslot),
+                    Joiners.equal(lambda lesson: lesson.studentGroup)
+                ]) \
+            .penalize("Student group conflict", HardSoftScore.ONE_HARD)
+----
+
+The `@constraint_provider` decorator allows OptaPlanner to use a function as a constraint provider.
+The function must take exactly one argument; the passed argument is a `ConstraintFactory` used for creating constraints.
+For more information, see https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintStreams[Constraint Streams in the OptaPlanner Docs].
+
+== Gather the domain objects in a planning solution
+
+A TimeTable wraps all Timeslot, Room, and Lesson instances of a single dataset.
+Furthermore, because it contains all lessons, each with a specific planning variable state, it is a planning solution and it has a score:
```

Comment:
```suggestion
Furthermore, because it contains all lessons, each with a specific planning variable state, the TimeTable class is a planning solution and it has a score:
```